WIP: split `Client` into `UnauthenticatedClient` and `Client`

dario23 opened the pull request:
a first attempt at the split between the unauthenticated (but connected) and the logged in state on the type level, as discussed in the comments of issue #80 . the only way to get a (logged in) `Client` instance is through successful calls to the `login`/`authenticate` methods. implementation-wise there's now a third type, `InnerClient`, which contains all lower-level functionality used by both `UnauthenticatedClient` and `Client` (to avoid duplicating the code or having weird cross-references between the `Client` and `UnauthenticatedClient` types.

as i'm mostly looking for feedback on the general direction of how i implemented the idea, there's not enough docs and i haven't ran `rustfmt` or `cargo test` yet etc. i'm especially hoping for feedback on

- do we want `connect`/`secure_connect` to stay in the `Client` type (but return an `UnauthenticatedClient`), so people can continue to write `Client::connect(..)`, or is it better to make the API break explicit here, as you have to deal with the return value of connect (the `UnauthenticatedClient`) differently now (i.e. don't provide a false sense of API continuity)?
- perhaps we might even want to re-add wrappers for the lower-level functions in `InnerClient` to the `Client` type, so we could reverse the changes to the (non-involved) `IdleHandle`. for now i only added wrappers for those that were marked `pub`.

i'm always happy to change things and re-submit. also i wasn't sure whether `UnauthenticatedClient`/`Client` or e.g. `Client`/`AuthenticatedClient` is the better naming; if people have thoughts on that i'm happy to hear those, too (in this first attempt i opted for the former, as this keeps more of the interface on the existing `Client` type).

jonhoo commented on src/client.rs:
```diff
@@ -218,85 +231,260 @@ impl Client<TcpStream> {
+        match self.inner.run_command(&format!("AUTHENTICATE {}", auth_type)) {
+            Ok(_) => self.do_auth_handshake(authenticator),
+            Err(e) => Err((e, self)),
+        }
```
Given the repetition of this, you may want to add a simple macro for it.

jonhoo commented:
I think this looks like a good approach! I'd suggest the following for naming/method placement:

 - `s/InnerClient/Connection/`
 - `s/Client/Session/`
 - `s/UnauthenticatedClient/Client/`
 - Add a free-standing `connect` method that returns a `Client`.

We could go even further and rename `UnauthenticatedClient` to `Builder`, where the `build` method is essentially `login`, but that might be too un-idiomatic.

The nicest way to work around all the `.inner.` things that have to be added is probably to add `impl Deref` and `impl DerefMut` for `Session` (and `Client`) with `Target = Connection`.

jonhoo commented on src/client.rs:
```diff
@@ -327,13 +359,14 @@ impl<T: Read + Write> UnauthenticatedClient<T> {
         username: &str,
         password: &str
-    ) -> ::std::result::Result<Client<T>, (Error, UnauthenticatedClient<T>)> {
+    ) -> ::std::result::Result<Session<T>, (Error, Client<T>)> {
```
Probably make `Client<T>` here just `Self`?

dario23 commented:
@jonhoo changed the naming of the various structs, implemented your suggestions on `.inner`, i.e. the `Deref` instances, and changed the tests accordingly, so they should now all pass as well.

depending on your preference i can squash some of the commits into one to avoid e.g. one iteration back and forth of adding `.inner` everywhere just to remove them again; or otherwise just leave all the commits like they are, to more accurately reflect the history/discussion here.

jonhoo commented:
Re 515d5747425257c5b8384b94eb15ba524802a19a: No, sadly not. There's the `#[doc]` attribute for including an external file, but you can't just straight include a Rust example file that way sadly :/

jonhoo commented on src/client.rs:
```diff
@@ -254,55 +310,212 @@ impl<T: Read + Write> Client<T> {
+        username: &str,
+        password: &str
+    ) -> ::std::result::Result<Session<T>, (Error, Client<T>)> {
+        // note that we need the explicit match blocks here for two reasons:
```
Should this comment now go away? Or perhaps be included in the macro?

dario23 replied:
ah, yes.. next commit (also including the comments from above) soon-ish :-)

dario23 commented:
docs (including moving the comment to the macro definition) and some examples added.

> I also have a somewhat silly request that you can ignore if you don't think it's worth it

i've been on both sides of very similar requests before, so happily included this one as well :-)

jonhoo commented:
Great, almost there! I think the comment still needs to be removed from `login`, but apart from that this PR is now basically mergeable :D

I'll merge once that's fixed, though @mattnenterprise will need to do a (major) release. Hopefully he'll also fix #69 when he's at it.

dario23 commented:
> I think the comment still needs to be removed from `login`

ah, right. pushed.

jonhoo commented:
Thank you!
